File: sample_source_code/python/src/math_utils.py
"""Math utility functions used across modules."""
from __future__ import annotations

import math


class MathUtils:
    """Pure static/class-method utilities. No instantiation needed."""

    @staticmethod
    def clamp(value: float, min_val: float, max_val: float) -> float:
        return max(min_val, min(max_val, value))

    @staticmethod
    def linear_decay(current: float, max_val: float, max_penalty: float) -> float:
        """Returns how much to subtract given current progress toward max_val."""
        if max_val <= 0:
            return 0.0
        return MathUtils.clamp((current / max_val) * max_penalty, 0.0, max_penalty)

    @staticmethod
    def fibonacci(n: int) -> int:
        """Memoised recursive Fibonacci."""
        memo: dict[int, int] = {}

        def _fib(k: int) -> int:
            if k <= 1:
                return k
            if k in memo:
                return memo[k]
            memo[k] = _fib(k - 1) + _fib(k - 2)
            return memo[k]

        return _fib(n)

    @staticmethod
    def standard_deviation(values: list[float]) -> float:
        """Population standard deviation."""
        if not values:
            return 0.0
        mean     = sum(values) / len(values)
        variance = sum((v - mean) ** 2 for v in values) / len(values)
        return math.sqrt(variance)

    @staticmethod
    def factorial(n: int) -> int:
        """Iterative factorial."""
        if n < 0:
            raise ValueError("Negative input to factorial")
        result = 1
        for i in range(2, n + 1):
            result *= i
        return result

File: sample_source_code/typescript/src/MathUtils.ts
/** Math utilities used across the codebase. */
export class MathUtils {
    static clamp(value: number, min: number, max: number): number {
        return Math.max(min, Math.min(max, value));
    }

    /** Returns how much to subtract given current progress toward max. */
    static linearDecay(current: number, max: number, maxPenalty: number): number {
        if (max <= 0) return 0;
        return MathUtils.clamp((current / max) * maxPenalty, 0, maxPenalty);
    }

    /** Memoised recursive Fibonacci. */
    static fibonacci(n: number): number {
        const memo = new Map<number, number>();
        function fib(k: number): number {
            if (k <= 1) return k;
            if (memo.has(k)) return memo.get(k)!;
            const r = fib(k - 1) + fib(k - 2);
            memo.set(k, r);
            return r;
        }
        return fib(n);
    }

    /** Population standard deviation. */
    static standardDeviation(values: number[]): number {
        if (!values.length) return 0;
        const mean     = values.reduce((s, v) => s + v, 0) / values.length;
        const variance = values.reduce((s, v) => s + (v - mean) ** 2, 0) / values.length;
        return Math.sqrt(variance);
    }

    /** Iterative factorial. */
    static factorial(n: number): number {
        if (n < 0) throw new Error('Negative input to factorial');
        let result = 1;
        for (let i = 2; i <= n; i++) result *= i;
        return result;
    }
}
